fe/fe-core/src/main/java/com/starrocks/connector/metastore/CachingMetastore.java

```diff
@@ -0,0 +1,108 @@
+    protected static final long NEVER_REFRESH = -1;
+
+    // Used to synchronize the refreshTable process
+    public final Map<DatabaseTableName, String> tableNameLockMap;
+    protected LoadingCache<String, List<String>> databaseNamesCache;
+    protected LoadingCache<String, List<String>> tableNamesCache;
+    protected LoadingCache<String, Database> databaseCache;
+    protected LoadingCache<DatabaseTableName, MetastoreTable> metastoreTableCache;
+    protected LoadingCache<DatabaseTableName, Table> tableCache;
+
+    public CachingMetastore(Executor executor, long expireAfterWriteSec, long refreshIntervalSec, long maxSize) {
+        databaseNamesCache = newCacheBuilder(NEVER_CACHE, NEVER_CACHE, NEVER_CACHE)
+                .build(asyncReloading(CacheLoader.from(this::loadAllDatabaseNames), executor));
+        tableNamesCache = newCacheBuilder(NEVER_CACHE, NEVER_CACHE, NEVER_CACHE)
+                .build(asyncReloading(CacheLoader.from(this::loadAllTableNames), executor));
+        databaseCache = newCacheBuilder(expireAfterWriteSec, refreshIntervalSec, maxSize)
+                .build(asyncReloading(CacheLoader.from(this::loadDb), executor));
+        metastoreTableCache = newCacheBuilder(expireAfterWriteSec, refreshIntervalSec, maxSize)
+                .build(asyncReloading(CacheLoader.from(this::loadMetastoreTable), executor));
+        tableCache = newCacheBuilder(expireAfterWriteSec, refreshIntervalSec, maxSize)
+                .build(asyncReloading(CacheLoader.from(this::loadTable), executor));
+
+        this.tableNameLockMap = Maps.newConcurrentMap();
+    }
+
+    protected static CacheBuilder<Object, Object> newCacheBuilder(long expiresAfterWriteSec, long refreshSec, long maximumSize) {
+        CacheBuilder<Object, Object> cacheBuilder = CacheBuilder.newBuilder();
+        if (expiresAfterWriteSec >= 0) {
+            cacheBuilder.expireAfterWrite(expiresAfterWriteSec, SECONDS);
+        }
+
+        if (refreshSec > 0 && expiresAfterWriteSec > refreshSec) {
+            cacheBuilder.refreshAfterWrite(refreshSec, SECONDS);
+        }
+
+        cacheBuilder.maximumSize(maximumSize);
+        return cacheBuilder;
+    }
+
+    protected abstract List<String> loadAllDatabaseNames();
+
+    protected abstract List<String> loadAllTableNames(String dbName);
+
+    protected abstract Database loadDb(String dbName);
+
+    protected abstract Table loadTable(DatabaseTableName databaseTableName);
+
+    protected abstract MetastoreTable loadMetastoreTable(DatabaseTableName databaseTableNam);
+
+    protected static <K, V> V get(LoadingCache<K, V> cache, K key) {
+        try {
+            return cache.getUnchecked(key);
+        } catch (UncheckedExecutionException e) {
+            LOG.error("Error occurred when loading cache", e);
+            throwIfInstanceOf(e.getCause(), StarRocksConnectorException.class);
+            throw e;
+        }
+    }
+
+    public synchronized void invalidateAll() {
```

Comment:
Do we need to add `invalidateTable` or `invalidateDatabase` to this class?

Reply:
we can add these if we need later